### PR TITLE
add backup option when receiving a client reset from Sync protocol

### LINKED ISSUE #3991: Support Client Reset
We should add better support for the server reseting the server state (= Client Reset).
Work is ongoing here: https://github.com/realm/realm-cocoa/pull/4507 and here https://github.com/realm/realm-object-store/pull/259

Once merged and we have finished #3742 we can implement the java equivalent to https://github.com/realm/realm-cocoa/pull/4507.

From a quick glance it looks like the following is enough:
- [ ] Extend `ObjectServerError` to include the userInfo dictionary.
- [ ] Add unit tests for it.

### PR DESCRIPTION
**Pull request by nhachicha**
mostly a copy/past from https://github.com/realm/realm-object-store/blob/ca5e3b8b7f8332a69970f226728889d459fbdd60/src/sync/sync_session.cpp#L352-L372 to add a backup operation when receiving a client_reset code 

Closes #3991

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/main/cpp/objectserver_shared.hpp [hunk 1]
```diff
@@ -61,27 +63,58 @@ class JniSession {
...
+            if (error_code.category() != realm::sync::protocol_error_category() &&
                     error_code.category() != realm::sync::client_error_category()) {
                 // FIXME: Consider below when moving to the OS sync manager.
                 // Ignore this error since it may cause exceptions in java ErrorCode.fromInt(). Throwing exception there
                 // will trigger "called with pending exception" later since the thread is created by java, and the
                 // endless loop is in native code. The java exception will never be thrown because of the endless loop
                 // will never quit to java land.
-                realm::jni_util::Log::e("Unhandled sync client error code %1, %2. is_fatal: %3.",
-                                        error_code.value(), error_code.message(), is_fatal);
+                realm::jni_util::Log::e("Unhandled sync client error code %1, %2. is_fatal:", error_code.value(), error_code.message(), is_fatal);
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/objectserver_shared.hpp]: Why remove `%3`?

File: realm/realm-library/src/main/cpp/objectserver_shared.hpp [hunk 2]
```diff
@@ -61,27 +63,58 @@ class JniSession {
...
+                                                  original_path,
+                                                  nullptr,
+                                                  nullptr,
+                                                  realm::util::Optional<std::string>(
+                                                          std::move(recovery_path)));
                 });
+            } else {
+                auto session_ref = weak_session_ref.lock();
+                if (session_ref) {
+
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/objectserver_shared.hpp]: Please remove empty line

File: realm/realm-library/src/main/cpp/objectserver_shared.hpp [hunk 3]
```diff
@@ -61,27 +63,58 @@ class JniSession {
...
+                        recovery_path = std::move(recovery_path)](const auto &manager) {
+                    realm::SyncFileActionMetadata(manager,
+                                                  realm::SyncFileActionMetadata::Action::HandleRealmForClientReset,
+                                                  original_path,
+                                                  nullptr,
+                                                  nullptr,
+                                                  realm::util::Optional<std::string>(
+                                                          std::move(recovery_path)));
                 });
+            } else {
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/objectserver_shared.hpp]: Please move `else {` to new line.

File: realm/realm-library/src/main/cpp/objectserver_shared.hpp [hunk 4]
```diff
@@ -61,27 +63,58 @@ class JniSession {
...
+
+                // Add a SyncFileActionMetadata marking the Realm as needing to be deleted.
+                auto recovery_path = realm::util::reserve_unique_file_name(
+                        realm::SyncManager::shared().recovery_directory_path(),
+                        realm::util::create_timestamped_template("recovered_realm"));
+                auto original_path = local_realm_path;
+
+                realm::SyncManager::shared().perform_metadata_update([original_path = std::move(original_path),
+                        recovery_path = std::move(recovery_path)](const auto &manager) {
+                    realm::SyncFileActionMetadata(manager,
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/objectserver_shared.hpp]: How does this work if the Realm is open? Doesn't this delete the underlying file?

Comment (kneth) [inline realm/realm-library/src/main/cpp/objectserver_shared.hpp]: It writes the action to the metadata table, and when the app restarts, the action will be performed. To do the reset here, `immediate_run_file_actions()` should be called.

File: realm/realm-library/src/main/cpp/objectserver_shared.hpp (hunk 1 above)
Comment (nhachicha) [inline realm/realm-library/src/main/cpp/objectserver_shared.hpp]: sorry typo, I was logging other things :p

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectserver/BoundState.java [hunk 5]
```diff
@@ -52,7 +52,22 @@ public void onError(ObjectServerError error) {
...
+            //  still consider this token to be valid (based on timestamps for example)
+            //
+            //  this may cause the server to send a fatal error (203 bad refresh) if we try to bind
+            //  the session with this token. To be safe we remove the token that has been considered by the
+            //  the server to be invalid.
+            session.getUser().removeAccessToken(session.getServerUrl());
+
+            // stop the session to avoid sending a bind to the server which will cause it to return
+            // a fatal 203 (bad refresh)
+            session.stopNativeSession();
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/internal/objectserver/BoundState.java]: This change seems to be same as from. I'll assume the review changes in https://github.com/realm/realm-java/pull/4040 will override this?

File: realm/realm-library/src/objectServer/java/io/realm/SyncUser.java [hunk 6]
```diff
@@ -392,6 +392,10 @@ private static String getManagementRealmUrl(URL authUrl) {
         }
     }
 
+    public void removeAccessToken (URI serverURL) {
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncUser.java]: So this one depends on #4040 ? You will have conflicts here after that merged. Just a reminder :)

**Pull request merged**